homeassistant/components/kef/media_player.py

```diff
@@ -0,0 +1,270 @@
+    speaker_type = config[CONF_TYPE]
+    port = config.get(CONF_PORT)
+    name = config.get(CONF_NAME)
+    maximum_volume = config.get(CONF_MAX_VOLUME)
+    volume_step = config.get(CONF_VOLUME_STEP)
+    inverse_speaker_mode = config.get(CONF_INVERSE_SPEAKER_MODE)
+    standby_time = config.get(CONF_STANDBY_TIME)
+
+    sources = SOURCES[speaker_type]
+
+    _LOGGER.debug(
+        "Setting up %s with host: %s, port: %s, name: %s, sources: %s",
+        DOMAIN,
+        host,
+        port,
+        name,
+        sources,
+    )
+
+    media_player = KefMediaPlayer(
+        name,
+        host,
+        port,
+        maximum_volume,
+        volume_step,
+        standby_time,
+        inverse_speaker_mode,
+        sources,
+        ioloop=hass.loop,
+    )
+    unique_id = media_player.unique_id
+    if unique_id in hass.data[DOMAIN]:
+        _LOGGER.debug("%s is already configured.", unique_id)
+    else:
+        hass.data[DOMAIN][unique_id] = media_player
+        async_add_entities([media_player], update_before_add=True)
+
+
+class KefMediaPlayer(MediaPlayerDevice):
+    """Kef Player Object."""
+
+    def __init__(
+        self,
+        name,
+        host,
+        port,
+        maximum_volume,
+        volume_step,
+        standby_time,
+        inverse_speaker_mode,
+        sources,
+        ioloop,
+    ):
+        """Initialize the media player."""
+        self._name = name
+        self._sources = sources
+        self._speaker = AsyncKefSpeaker(
+            host,
+            port,
+            volume_step,
+            maximum_volume,
+            standby_time,
+            inverse_speaker_mode,
+            ioloop=ioloop,
+        )
+
+        self._state = STATE_UNKNOWN
+        self._muted = None
+        self._source = None
+        self._volume = None
+        self._is_online = None
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        return self._state
+
+    async def async_update(self):
+        """Update latest state."""
+        _LOGGER.debug("Running async_update")
+        try:
+            self._is_online = await self._speaker.is_online()
+            if self._is_online:
+                (
+                    self._volume,
+                    self._muted,
+                ) = await self._speaker.get_volume_and_is_muted()
+                state = await self._speaker.get_state()
+                self._source = state.source
+                self._state = STATE_ON if state.is_on else STATE_OFF
+            else:
+                self._muted = None
+                self._source = None
+                self._volume = None
+                self._state = STATE_OFF
+        except (ConnectionRefusedError, ConnectionError, TimeoutError) as err:
+            _LOGGER.debug("Error in `update`: %s", err)
+            self._state = STATE_UNKNOWN
```

Comment:
Use `None` to represent unknown state.

Reply:
Fixed in 4c87e805debe5576978cf62944b3afde4809916e.